Assign multiple rotations per application

oyving opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

Commit:
Add support for EndpointId in EndpointBuilder

Commit:
Update javadoc to explain new behavior on default endpoint ID

Commit:
Do not allow duplicate IDs of endpoints when reading deployments.xml

Commit:
Add support for endpoints in APB

Commit:
Remove "support" for duplicate endpoint IDs in ApplicationSerializer

Commit:
Put information on assigned regions in AssignedRotation class

Commit:
Add methods to get rotations based on ID, and to assign multiple rotations per app

Commit:
Take endpoints into account when generating EndpointList

Commit:
Use EndpointId to generate Endpoints in EndpointList

Commit:
Assign multiple rotations per application during deploy

Use updated logic to generate endpoint names, use the new interface to send
ContainerEndpoints to the configserver, and use new method to assign multiple
rotations per application.

Commit:
Change signature from List to Set so we don't respect ordering too much

Commit:
Update tests to take into account rotation changes

- We no longer generate legacy endpoint names
- We can have multiple global endpoints per application
- We only generate one global endpoint per endpoint id

Commit:
Merge remote-tracking branch 'origin/master' into ogronnesby/assign-multiple-rotations

Conflicts:
	controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/configserver/ConfigServer.java
	controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java
	controller-server/src/test/java/com/yahoo/vespa/hosted/controller/integration/ConfigServerMock.java
	controller-server/src/test/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializerTest.java

mpolden commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/rotation/RotationRepository.java:
```diff
@@ -75,6 +88,57 @@ public Rotation getOrAssignRotation(Application application, RotationLock lock)
     }
 
+    public List<AssignedRotation> getOrAssignRotations(Application application, RotationLock lock) {
+        if (application.deploymentSpec().globalServiceId().isPresent() && ! application.deploymentSpec().endpoints().isEmpty()) {
```
`DeploymentSpec` already enforces this.

oyving replied:
I agree, but I think it doesn't hurt to enforce it here as well?

mpolden commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/rotation/RotationRepository.java:
```diff
@@ -75,6 +88,57 @@ public Rotation getOrAssignRotation(Application application, RotationLock lock)
         return findAvailableRotation(application, lock);
     }
 
+    public List<AssignedRotation> getOrAssignRotations(Application application, RotationLock lock) {
```
Please add javadoc.

mpolden commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/EndpointList.java:
```diff
@@ -67,16 +67,14 @@ public static EndpointList of(Stream<Endpoint> endpoints) {
-    public static EndpointList defaultGlobal(ApplicationId application, SystemName system) {
-        // Rotation name is always default in the routing layer
-        RotationName rotation = RotationName.from("default");
+    public static EndpointList defaultGlobal(ApplicationId application, EndpointId endpointId, SystemName system) {
```
This should be renamed since "default" is no longer implied. Just `create` perhaps?

Commit:
Add convenience method to create AssignedRotations

Commit:
Fix issue with rotations changing after assignment

- Make sure we only include production zones in the assignment
- Always run deployment through rotation assignment logic, so regions are cleared
  from persisted assignments even if there are no remaining rotations in the configuration.
- Add tests to expose this logic.
- Add tiny convenience method for AssignedRotation to create them in tests.

Commit:
Split finding needed rotations and assigning them to two steps

- We find the needed rotations and check if we have enough before assigning,
  throwing a sensible exception if there are not enough.

- Moved where we find available rotations so we are certain we have the lock
  during rotation assignment (this was guaranteed before, but now it is more
  explicit in the signature).

- Added some javadoc to the new assignment method.

Commit:
Rename .defaultGlobal => .create

Commit:
Introduce feature flag on application ID for new endpoints behavior

- Added a feature flag on application level for new endpoints behavior
- Reintroduced a test of legacy behavior to validate old logic
